Add support for simple tooltips

pythongosssss opened the pull request:
Adds basic support for a TOOLTIPS property, these show in the UI when hovering inputs/outputs/widgets.
DOM widgets will the use the native browser tooltips.

There is a setting to disable these.

I've added samples to CLIPTextEncode and LoraLoader, with the basic descriptions taken (and slightly tweaked) from
https://blenderneko.github.io/ComfyUI-docs/ (CC-BY-SA)
For a starting point we could either extract the rest from there, or from https://docs.getsalt.ai/md/Comfy/Nodes/LoraLoader/ however the descriptions here are a lot more verbose.

![tooltip](https://github.com/comfyanonymous/ComfyUI/assets/125205205/980c6879-53f2-4a99-b46b-c6a614cc9770)

Bocian-1 commented:
Would it be possible to have the ability to set custom tooltips on group nodes from the group node management menu?
Not necessarily per widget, just the group node itself would be great.

DrJKL commented on web/style.css:
```diff
@@ -557,3 +557,20 @@ dialog::backdrop {
+	box-shadow: 0 0 5px rgba(0, 0, 0, 0.4);
+	font-family: sans-serif;
+	white-space: pre-wrap;
+}
```
```suggestion
}

```

DrJKL commented on web/extensions/core/tooltip.css:
```diff
@@ -0,0 +1,16 @@
+	box-shadow: 0 0 5px rgba(0, 0, 0, 0.4);
+	font-family: sans-serif;
+	white-space: pre-wrap;
+}
```
```suggestion
}

```

DrJKL reviewed (commented):
Trailing newlines help in the diffs in future changes.

shawnington commented:
This is how tool tips should be for comfy. Standardized format, simple and descriptive of what things do.

pythongosssss commented:
I've added tooltips to the nodes that are included in the default workflow, some feedback on the wording/accuracy of them would be appreciated (especially the KSampler) trying to be correct but still clear for the average user

DrJKL commented on web/style.css:
```diff
@@ -636,3 +636,20 @@ dialog::backdrop {
 }
+
+.comfy-graph-tooltip {
+	z-index: 99999;
```
Same here 🙏🏻

DrJKL commented on web/style.css:
```diff
@@ -636,3 +636,20 @@ dialog::backdrop {
 	display: none;
 }
+
+.comfy-graph-tooltip {
```
Also, why the styles in 2 places?

pythongosssss replied:
Deleted file

pythongosssss commented on nodes.py:
```diff
@@ -510,6 +524,12 @@ def INPUT_TYPES(s):
+        "input": { 
+            "ckpt_name": "The name of the checkpoint (model) to load.",
+        },
+        "output": ("The model used for denoising latents.", "The CLIP model used for encoding text prompts.", "The VAE model used for encoding and decoding images to and from latent space.")
```
Yeah, something to consider in future would be a redesign of the node class converting outputs into a structure more similar to the inputs would be nice

pythongosssss commented on nodes.py:
```diff
@@ -590,6 +610,16 @@ def INPUT_TYPES(s):
+            "model": "The diffusion model the LoRA will be applied to.",
+            "clip": "The CLIP model the LoRA will be applied to.",
+            "lora_name": "The name of the LoRA.",
+            "strength_model": "How strongly to modify the diffusion model. This value can be negative.",
```
Recommended weight for loras varies significantly, especially with the "slider" ones, so its going to generally be "check the lora description!"

pythongosssss commented on nodes.py:
```diff
@@ -1029,6 +1059,14 @@ def INPUT_TYPES(s):
     CATEGORY = "latent"
+    TOOLTIPS = {
+        "input": { 
+            "width": "The width of the latent images in pixels.",
```
It should enforce a step of 8 so not sure required here (also, some custom nodes require 64 here)

pythongosssss commented on web/extensions/core/tooltip.css:
```diff
@@ -0,0 +1,16 @@
+.comfy-graph-tooltip {
+	z-index: 99999;
```
Deleted file

pythongosssss commented on nodes.py:
```diff
@@ -1029,6 +1059,14 @@ def INPUT_TYPES(s):
+        "input": { 
+            "width": "The width of the latent images in pixels.",
+            "height": "The height of the latent images in pixels.",
+            "batch_size": "The number of latent images in the batch.",
```
Yes 4096, but I dont think it's something that would need to go in the tooltip as the interface capping the value should be clear enough

comfyanonymous commented:
Is this good to merge?

Amorano commented:
I know you are tired of hearing about my versions of system, but what I am asking here is because it is important to the ecosystem, not my ego.

Is there no way to put the data for the tooltip in place where the actual input/output is?

![image](https://github.com/user-attachments/assets/b9c2372b-0352-4858-a57d-ff1ca2ecb5f3)

Is there a specific reason the data is disconnected from the meta definition it is about?

The same for the outputs, they should be defined in the same meta data bucket:

"required": {
"optional": {
"output": {

with all the same supports of the other fields i.e. tooltips.

huchenlei commented:
> I know you are tired of hearing about my versions of system, but what I am asking here is because it is important to the ecosystem, not my ego.
> 
> Is there no way to put the data for the tooltip in place where the actual input/output is?
> 
> ![image](https://private-user-images.githubusercontent.com/225178/350455368-b9c2372b-0352-4858-a57d-ff1ca2ecb5f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0MTkxNjIsIm5iZiI6MTcyMTQxODg2MiwicGF0aCI6Ii8yMjUxNzgvMzUwNDU1MzY4LWI5YzIzNzJiLTAzNTItNDg1OC1hNTdkLWZmMWNhMmVjYjVmMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOVQxOTU0MjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMmE5ZmVhMTRjODExMjgzZWYzOTRkMjk4Mjk2NzY2YWQ4OTVkYzkzYjA4NzY2Mzc4NzIxN2Q4MjA5NjYwNjE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.E4BnFjVWPyZLwyr9SHbqb8nEtrGlSkjxvYBsK6vKzs8)
> 
> Is there a specific reason the data is disconnected from the meta definition it is about?
> 
> The same for the outputs, they should be defined in the same meta data bucket:
> 
> "required": { "optional": { "output": {
> 
> with all the same supports of the other fields i.e. tooltips.

+1 on this

huchenlei commented:
I think even the output does not have a dict format right now, we can still do the input part and make output tooltip as an array of items just like other output fields. This is ugly but more aligned mentally with the current system.

Note:
In the new frontend we already has class-transformer translating the output to a good format. Here is a sample test case taken from the frontend repo:
```js
it('should handle multiple outputs including COMBO type', () => {
    const plainObject = {
      name: 'MultiOutputNode',
      display_name: 'Multi Output Node',
      category: 'Advanced',
      python_module: 'advanced_module',
      description: 'A node with multiple outputs',
      input: {},
      output: ['STRING', ['COMBO', 'option1', 'option2'], 'FLOAT'],
      output_is_list: [true, false, false],
      output_name: ['stringOutput', 'comboOutput', 'floatOutput']
    }

    const result = plainToClass(ComfyNodeDefImpl, plainObject)

    expect(result.output.all).toEqual([
      {
        index: 0,
        name: 'stringOutput',
        type: 'STRING',
        is_list: true
      },
      {
        index: 1,
        name: 'comboOutput',
        type: 'COMBO',
        is_list: false,
        comboOptions: ['COMBO', 'option1', 'option2']
      },
      {
        index: 2,
        name: 'floatOutput',
        type: 'FLOAT',
        is_list: false
      }
    ])
  })
```

pythongosssss commented:
* Moved tooltips to OUTPUT_TOOLTIPS and the inputs dict
* Added support for node level tooltip (hover the title) using node description
* Added node descriptions for default workflow
* Reworked tooltip checking for better performance
* Add tooltips for control_* widgets

BuffMcBigHuge commented:
@pythongosssss 

Tooltips are no longer working. Mouseover doesn't show them.

![image](https://github.com/user-attachments/assets/5cd677bd-ff8c-4f62-a38a-2ee37b602c75)

![image](https://github.com/user-attachments/assets/fa8d0161-a16c-40b7-ad17-f0743d4081ff)
